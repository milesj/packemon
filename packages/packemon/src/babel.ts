import fs from 'fs-extra';
import { TransformOptions as ConfigStructure } from '@babel/core';
import { Path, toArray } from '@boost/common';
import { FeatureFlags } from './types';
import {
	CodeArtifact,
	DEFAULT_SUPPORT,
	Format,
	getBabelInputConfig,
	getBabelOutputConfig,
	Package,
	PackemonPackage,
	Platform,
	Project,
	Support,
} from '.';

const format = (process.env.PACKEMON_FORMAT ?? 'lib') as Format;
const support = (process.env.PACKEMON_SUPPORT ?? DEFAULT_SUPPORT) as Support;
const project = new Project(process.cwd());

function getBabelConfig(artifact: CodeArtifact, featureFlags: FeatureFlags): ConfigStructure {
	const inputConfig = getBabelInputConfig(artifact, featureFlags);
	const outputConfig = getBabelOutputConfig(
		artifact.platform,
		artifact.support,
		artifact.builds[0].format,
		featureFlags,
	);

	return {
		// Input must come first
		plugins: [...inputConfig.plugins!, ...outputConfig.plugins!],
		// Env must come first
		presets: [...outputConfig.presets!, ...inputConfig.presets!],
	};
}

export interface ConfigOptions {
	format?: Format;
	platform?: Platform;
	support?: Support;
}

export function createConfig(folder: string, options: ConfigOptions = {}): ConfigStructure {
	const path = new Path(folder);
	const contents = fs.readJsonSync(path.append('package.json').path()) as PackemonPackage;

	// Create package and configs
	const pkg = new Package(project, path, contents);

	if (pkg.packageJson.packemon) {
		pkg.setConfigs(toArray(pkg.packageJson.packemon));
	}

	// Determine the lowest platform to support
	const platforms = pkg.configs.map((config) => config.platform);
	let lowestPlatform: Platform = 'node';

	// istanbul ignore next
	if (platforms.includes('browser')) {
		lowestPlatform = 'browser';
	} else if (platforms.includes('native')) {
		lowestPlatform = 'native';
	}

	// Generate artifact and builds
	const artifact = new CodeArtifact(pkg, [{ format: options.format ?? format }]);
	artifact.bundle = false;
	artifact.platform = options.platform ?? lowestPlatform;
	artifact.support = options.support ?? support;

	return getBabelConfig(artifact, pkg.getFeatureFlags());
}

export function createRootConfig(options?: ConfigOptions): ConfigStructure {
	const config = createConfig(process.cwd(), options);

	return {
		...config,
		babelrc: true,
		babelrcRoots: project.getWorkspaceGlobs({ relative: true }),
		// Support React Native libraries by default
		overrides: [
			{
				presets: ['@babel/preset-flow'],
				test: /node_modules\/((jest-)?react-native|@react-native(-community)?)/iu,
			},
		],
	};
}
