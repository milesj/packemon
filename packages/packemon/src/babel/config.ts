import { PluginItem, TransformOptions as ConfigStructure } from '@babel/core';
import { CodeArtifact } from '../CodeArtifact';
import { BROWSER_TARGETS, NATIVE_TARGETS, NODE_SUPPORTED_VERSIONS } from '../constants';
import { FeatureFlags, Format, Platform, Support } from '../types';
import { resolve, resolveFromBabel } from './resolve';

// https://babeljs.io/docs/en/babel-preset-env
export interface PresetEnvOptions {
	browserslistEnv?: string;
	bugfixes?: boolean;
	corejs?: 2 | 3 | { version: 2 | 3; proposals: boolean };
	debug?: boolean;
	exclude?: string[];
	forceAllTransforms?: boolean;
	ignoreBrowserslistConfig?: boolean;
	include?: string[];
	loose?: boolean;
	modules?: 'amd' | 'auto' | 'cjs' | 'commonjs' | 'systemjs' | 'umd' | false;
	shippedProposals?: boolean;
	spec?: boolean;
	targets?: Record<string, string[] | string> | string[] | string;
	useBuiltIns?: 'entry' | 'usage' | false;
}

function getPlatformEnvOptions(
	platform: Platform,
	support: Support,
	format: Format,
): PresetEnvOptions {
	let modules: PresetEnvOptions['modules'] = false;

	if (format === 'umd') {
		modules = 'umd';
	} else if (format === 'cjs' || format === 'lib') {
		modules = 'cjs'; // Babel CommonJS
	}

	switch (platform) {
		case 'browser':
			return {
				modules,
				targets: { browsers: BROWSER_TARGETS[support] },
			};

		case 'native':
			return {
				modules,
				targets: { browsers: NATIVE_TARGETS[support] },
			};

		case 'node':
			return {
				// Async/await has been available since v7
				exclude: [
					'@babel/plugin-transform-regenerator',
					'@babel/plugin-transform-async-to-generator',
				],
				modules,
				targets: {
					node: NODE_SUPPORTED_VERSIONS[support],
				},
			};

		default:
			throw new Error(`Unknown platform "${platform}".`);
	}
}

function getSharedConfig(
	plugins: PluginItem[],
	presets: PluginItem[],
	features: FeatureFlags,
): ConfigStructure {
	return {
		caller: {
			name: 'packemon',
		},
		comments: true,
		parserOpts: {
			sourceType: 'unambiguous',
			strictMode: features.strict,
		},
		plugins,
		presets,
		// Do NOT load root `babel.config.js` as we need full control
		configFile: false,
		// Do load branch `.babelrc.js` files for granular customization
		babelrc: true,
		babelrcRoots: features.workspaces,
	};
}

// The input config should only parse special syntax, not transform and downlevel.
// This applies to all formats within a build target.
export function getBabelInputConfig(
	artifact: CodeArtifact,
	features: FeatureFlags,
): Omit<ConfigStructure, 'exclude' | 'include'> {
	const plugins: PluginItem[] = [];
	const presets: PluginItem[] = [];

	if (features.flow) {
		presets.push([resolve('@babel/preset-flow'), { allowDeclareFields: true }]);
	}

	if (features.typescript) {
		presets.push([resolve('@babel/preset-typescript'), { allowDeclareFields: true }]);

		// When decorators are used, class properties must be loose
		if (features.decorators) {
			plugins.push(
				[resolve('@babel/plugin-proposal-decorators'), { legacy: true }],
				[resolveFromBabel('@babel/plugin-proposal-class-properties'), { loose: true }],
				[resolveFromBabel('@babel/plugin-proposal-private-methods'), { loose: true }],
				[resolveFromBabel('@babel/plugin-proposal-private-property-in-object'), { loose: true }],
			);
		}
	}

	if (features.react) {
		presets.push([
			resolve('@babel/preset-react'),
			{
				development: __DEV__,
				throwIfNamespace: true,
			},
		]);
	}

	return getSharedConfig(plugins, presets, features);
}

// The output config does all the transformation and downleveling through the preset-env.
// This is handled per output since we need to configure based on target + format combinations.
export function getBabelOutputConfig(
	platform: Platform,
	support: Support,
	format: Format,
	features: FeatureFlags,
): ConfigStructure {
	const plugins: PluginItem[] = [];
	const presets: PluginItem[] = [];
	const isESM = format === 'esm' || format === 'mjs';

	// ENVIRONMENT

	const envOptions: PresetEnvOptions = {
		// Prefer spec compliance in development
		spec: __DEV__,
		loose: false,
		// Consumers must polyfill accordingly
		useBuiltIns: false,
		// Transform features accordingly
		bugfixes: true,
		shippedProposals: true,
		// Platform specific
		...getPlatformEnvOptions(platform, support, format),
	};

	presets.push([resolve('@babel/preset-env'), envOptions]);

	// PLUGINS

	if (platform === 'browser' || platform === 'native') {
		// While modern browsers support these features, Node.js <= 12 does not,
		// which results in failing builds trying to parse the syntax.
		// Let's only apply this for the lib format, but allow it for esm.
		if (format === 'lib') {
			plugins.push(
				resolveFromBabel('@babel/plugin-proposal-logical-assignment-operators'),
				resolveFromBabel('@babel/plugin-proposal-nullish-coalescing-operator'),
				resolveFromBabel('@babel/plugin-proposal-optional-chaining'),
			);
		}

		// Both browsers and Node.js support these features outside of legacy targets
		if (support === 'legacy') {
			plugins.push(
				[
					resolve('babel-plugin-transform-async-to-promises'),
					{ inlineHelpers: true, target: 'es5' },
				],
				[
					resolve('@babel/plugin-transform-runtime'),
					{ helpers: false, regenerator: true, useESModules: isESM },
				],
			);
		}
	} else {
		// Use `Object.assign` when available
		plugins.push(
			[resolveFromBabel('@babel/plugin-transform-destructuring'), { useBuiltIns: true }],
			[resolveFromBabel('@babel/plugin-proposal-object-rest-spread'), { useBuiltIns: true }],
		);
	}

	// Support our custom plugins
	if (platform === 'node') {
		plugins.push([resolve('babel-plugin-cjs-esm-interop'), { format: isESM ? 'mjs' : 'cjs' }]);
	}

	plugins.push(
		resolve('babel-plugin-conditional-invariant'),
		resolve('babel-plugin-env-constants'),
	);

	return getSharedConfig(plugins, presets, features);
}
