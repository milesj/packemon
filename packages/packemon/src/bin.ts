import { checkPackageOutdated, Program } from '@boost/cli';
import { getVersion } from './helpers/getVersion';
import {
	BuildCommand,
	CleanCommand,
	InitCommand,
	PackCommand,
	ScaffoldCommand,
	ValidateCommand,
	WatchCommand,
} from '.';

let version = '0.0.0-internal';

try {
	version = getVersion();
} catch {
	// Ignore
}

async function run() {
	const program = new Program({
		bin: 'packemon',
		footer: 'Documentation: https://packemon.dev',
		name: 'Packemon',
		version,
	});

	program
		.categories({
			filter: 'Filtering',
		})
		.middleware(checkPackageOutdated('packemon', version))
		.register(new BuildCommand())
		.register(new CleanCommand())
		.register(new InitCommand())
		.register(new PackCommand())
		.register(new ScaffoldCommand())
		.register(new ValidateCommand())
		.register(new WatchCommand());

	await program.runAndExit(process.argv);
}

void run();
