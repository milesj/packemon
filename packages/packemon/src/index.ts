/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './Artifact';
export * from './babel/config';
export * from './CodeArtifact';
export * from './commands/Base';
export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Init';
export * from './commands/Pack';
export * from './commands/Scaffold';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './Package';
export * from './PackageValidator';
export * from './Packemon';
export * from './Project';
export * from './rollup/config';
export * from './types';
export * from './TypesArtifact';
