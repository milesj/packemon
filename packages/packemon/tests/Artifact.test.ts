import execa from 'execa';
import fsx from 'fs-extra';
import { rollup } from 'rollup';
import { applyStyle } from '@boost/cli';
import { Path } from '@boost/common';
import { Artifact } from '../src/Artifact';
import { getRollupConfig } from '../src/rollup/config';
import { getFixturePath, loadPackageAtPath, mockSpy } from './helpers';

jest.mock('../src/rollup/config', () => ({
	getRollupConfig: jest.fn(() => ({
		input: true,
		output: [
			{ originalFormat: 'lib', a: true },
			{ originalFormat: 'cjs', b: true },
			{ originalFormat: 'mjs', c: true },
		],
	})),
}));

jest.mock('execa');

jest.mock('rimraf', () =>
	jest.fn((path, cb) => {
		cb();
	}),
);

jest.mock('rollup', () => ({ rollup: jest.fn() }));

class TestArtifact extends Artifact {
	log = this.logWithSource.bind(this);
}

describe('Artifact', () => {
	const fixturePath = new Path(getFixturePath('project'));
	let artifact: TestArtifact;

	beforeEach(() => {
		artifact = new TestArtifact(loadPackageAtPath(fixturePath), [
			{ format: 'cjs' },
			{ format: 'mjs' },
		]);
		artifact.inputs = { index: 'src/index.ts' };
	});

	it('returns label when cast to string', () => {
		expect(String(artifact)).toBe('node:stable:cjs,mjs');
	});

	describe('build()', () => {
		it('builds code and types', async () => {
			const codeSpy = jest.spyOn(artifact, 'buildCode').mockImplementation();
			const typesSpy = jest.spyOn(artifact, 'buildTypes').mockImplementation();

			await artifact.build({}, {}, {});

			expect(codeSpy).toHaveBeenCalled();
			expect(typesSpy).toHaveBeenCalled();
		});
	});

	describe('buildCode()', () => {
		let bundleWriteSpy: jest.SpyInstance;

		beforeEach(() => {
			bundleWriteSpy = jest.fn(() => ({ output: [{ type: 'chunk', code: 'code' }] }));

			mockSpy(rollup)
				.mockReset()
				.mockImplementation(() => ({
					cache: { cache: true },
					generate: bundleWriteSpy,
					write: bundleWriteSpy,
				}));

			artifact.builds.push({ format: 'lib' });
		});

		it('generates rollup config using input config', async () => {
			await artifact.buildCode({ typescript: true }, {});

			expect(getRollupConfig).toHaveBeenCalledWith(
				artifact,
				{ typescript: true },
				expect.any(Object),
			);
			expect(rollup).toHaveBeenCalledWith({
				input: true,
				onwarn: expect.any(Function),
			});
		});

		it('writes a bundle and stats for each build', async () => {
			await artifact.buildCode({}, {});

			expect(bundleWriteSpy).toHaveBeenCalledWith({ a: true });
			expect(artifact.builds[0].stats?.size).toBe(4);

			expect(bundleWriteSpy).toHaveBeenCalledWith({ b: true });
			expect(artifact.builds[1].stats?.size).toBe(4);

			expect(bundleWriteSpy).toHaveBeenCalledWith({ c: true });
			expect(artifact.builds[2].stats?.size).toBe(4);
		});
	});

	describe('buildTypes()', () => {
		it('doesnt run if typescript is disabled', async () => {
			await artifact.buildTypes({ typescript: false });

			expect(execa).not.toHaveBeenCalled();
		});

		it('doesnt run if a specific build has declarations disabled', async () => {
			artifact.builds[0].declaration = true;

			await artifact.buildTypes({ typescript: true });

			expect(execa).toHaveBeenCalledTimes(1);
		});

		it('runs `tsc` for each build', async () => {
			artifact.builds[0].declaration = true;
			artifact.builds[1].declaration = true;

			await artifact.buildTypes({ typescript: true });

			expect(execa).toHaveBeenCalledWith(
				'tsc',
				[
					'--declaration',
					'--declarationDir',
					'cjs',
					'--declarationMap',
					'--emitDeclarationOnly',
					'--project',
					'tsconfig.cjs.json',
				],
				{ cwd: fixturePath.path(), preferLocal: true },
			);

			expect(execa).toHaveBeenCalledWith(
				'tsc',
				[
					'--declaration',
					'--declarationDir',
					'mjs',
					'--declarationMap',
					'--emitDeclarationOnly',
					'--project',
					'tsconfig.mjs.json',
				],
				{ cwd: fixturePath.path(), preferLocal: true },
			);
		});

		it('runs `tsc` for each build in composite mode', async () => {
			artifact.builds[0].declaration = true;
			artifact.builds[1].declaration = true;

			await artifact.buildTypes({ typescript: true, typescriptComposite: true });

			expect(execa).toHaveBeenCalledWith('tsc', ['--build', '--force', 'tsconfig.cjs.json'], {
				cwd: fixturePath.path(),
				preferLocal: true,
			});

			expect(execa).toHaveBeenCalledWith('tsc', ['--build', '--force', 'tsconfig.mjs.json'], {
				cwd: fixturePath.path(),
				preferLocal: true,
			});
		});
	});

	describe('clean()', () => {
		it('removes the dir for each format', async () => {
			const spy = jest.spyOn(fsx, 'remove');

			await artifact.clean();

			expect(spy).toHaveBeenCalledWith(fixturePath.append('assets').path());
			expect(spy).toHaveBeenCalledWith(fixturePath.append('dts').path());
			expect(spy).toHaveBeenCalledWith(fixturePath.append('cjs').path());
			expect(spy).toHaveBeenCalledWith(fixturePath.append('mjs').path());
		});
	});

	describe('getBuildOutput()', () => {
		beforeEach(() => {
			artifact.platform = 'node';
		});

		it('returns metadata for `lib` format', () => {
			expect(artifact.getBuildOutput('lib', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'js',
				entryPath: './lib/index.js',
				folder: 'lib',
			});
		});

		it('returns metadata for `esm` format', () => {
			expect(artifact.getBuildOutput('esm', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'js',
				entryPath: './esm/index.js',
				folder: 'esm',
			});
		});

		it('returns metadata for `umd` format', () => {
			expect(artifact.getBuildOutput('umd', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'js',
				entryPath: './umd/index.js',
				folder: 'umd',
			});
		});

		it('returns metadata for `cjs` format', () => {
			expect(artifact.getBuildOutput('cjs', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'cjs',
				entryPath: './cjs/index.cjs',
				folder: 'cjs',
			});
		});

		it('returns metadata for `mjs` format', () => {
			expect(artifact.getBuildOutput('mjs', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'mjs',
				entryPath: './mjs/index.mjs',
				folder: 'mjs',
			});
		});

		describe('shared lib', () => {
			it('includes platform in folder when shared lib required', () => {
				artifact.sharedLib = true;

				expect(artifact.getBuildOutput('lib', 'index')).toEqual({
					declExt: undefined,
					declPath: undefined,
					entryExt: 'js',
					entryPath: './lib/node/index.js',
					folder: 'lib/node',
				});
			});

			it('ignores shared lib if not `lib` format', () => {
				artifact.sharedLib = true;

				expect(artifact.getBuildOutput('esm', 'index')).toEqual({
					declExt: undefined,
					declPath: undefined,
					entryExt: 'js',
					entryPath: './esm/index.js',
					folder: 'esm',
				});
			});
		});

		it('public api: uses original source file path', () => {
			artifact.api = 'public';
			artifact.inputs = { index: 'src/some/other/file.ts' };

			expect(artifact.getBuildOutput('mjs', 'index')).toEqual({
				declExt: undefined,
				declPath: undefined,
				entryExt: 'mjs',
				entryPath: './mjs/some/other/file.mjs',
				folder: 'mjs',
			});
		});

		describe('types', () => {
			it('returns declaration fields', () => {
				expect(artifact.getBuildOutput('lib', 'index', true)).toEqual({
					declExt: 'd.ts',
					declPath: './lib/index.d.ts',
					entryExt: 'js',
					entryPath: './lib/index.js',
					folder: 'lib',
				});
			});

			it('supports cts', () => {
				artifact.inputs = { index: 'src/index.cts' };
				artifact.builds = [{ format: 'cjs' }];

				expect(artifact.getBuildOutput('cjs', 'index', true)).toEqual({
					declExt: 'd.cts',
					declPath: './cjs/index.d.cts',
					entryExt: 'cjs',
					entryPath: './cjs/index.cjs',
					folder: 'cjs',
				});
			});

			it('supports mts', () => {
				artifact.inputs = { index: 'src/index.mts' };
				artifact.builds = [{ format: 'mjs' }];

				expect(artifact.getBuildOutput('mjs', 'index', true)).toEqual({
					declExt: 'd.mts',
					declPath: './mjs/index.d.mts',
					entryExt: 'mjs',
					entryPath: './mjs/index.mjs',
					folder: 'mjs',
				});
			});
		});
	});

	describe('getInputPaths()', () => {
		it('returns an absolute path for every input', () => {
			expect(artifact.getInputPaths()).toEqual({
				index: fixturePath.append('src/index.ts').path(),
			});
		});
	});

	describe('getPackageExports()', () => {
		beforeEach(() => {
			artifact.builds = [];
		});

		it('adds exports based on input file and output name', () => {
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: undefined,
						default: './lib/index.js',
					},
					default: './lib/index.js',
				},
			});
		});

		it('adds exports based on input file and output name when shared lib required', () => {
			artifact.sharedLib = true;
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: undefined,
						default: './lib/node/index.js',
					},
					default: './lib/node/index.js',
				},
			});
		});

		it('supports subpath file exports when output name is not "index"', () => {
			artifact.inputs = { sub: './src/sub.ts' };
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'./sub': {
					node: {
						types: undefined,
						default: './lib/sub.js',
					},
					default: './lib/sub.js',
				},
			});
		});

		it('supports conditional exports when there are multiple builds', () => {
			artifact.builds.push({ format: 'lib' }, { format: 'mjs' }, { format: 'cjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						import: {
							types: undefined,
							default: './mjs/index.mjs',
						},
						require: {
							types: undefined,
							default: './cjs/index.cjs',
						},
					},
					default: './lib/index.js',
				},
			});
		});

		it('supports conditional exports with types when there are multiple builds', () => {
			artifact.builds.push(
				{ declaration: true, format: 'lib' },
				{ declaration: true, format: 'mjs' },
				{ declaration: true, format: 'cjs' },
			);

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					types: './lib/index.d.ts',
					node: {
						import: {
							types: './mjs/index.d.ts',
							default: './mjs/index.mjs',
						},
						require: {
							types: './cjs/index.d.ts',
							default: './cjs/index.cjs',
						},
					},
					default: './lib/index.js',
				},
			});
		});

		it('skips `default` export when there is no `lib` build', () => {
			artifact.inputs = { sub: './src/sub.ts' };
			artifact.builds.push({ format: 'mjs' }, { format: 'cjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'./sub': {
					node: {
						import: {
							types: undefined,
							default: './mjs/sub.mjs',
						},
						require: {
							types: undefined,
							default: './cjs/sub.cjs',
						},
					},
				},
			});
		});

		it('changes export namespace to "browser" when a `browser` platform', () => {
			artifact.platform = 'browser';
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					browser: {
						types: undefined,
						default: './lib/index.js',
					},
					default: './lib/index.js',
				},
			});
		});

		it('changes export namespace to "react-native" when a `native` platform', () => {
			artifact.platform = 'native';
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					'react-native': {
						default: './lib/index.js',
					},
					default: './lib/index.js',
				},
			});
		});

		it('supports lib', () => {
			artifact.builds.push({ format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: undefined,
						default: './lib/index.js',
					},
					default: './lib/index.js',
				},
			});
		});

		it('supports lib with types', () => {
			artifact.builds.push({ declaration: true, format: 'lib' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					types: './lib/index.d.ts',
					node: {
						types: './lib/index.d.ts',
						default: './lib/index.js',
					},
					default: './lib/index.js',
				},
			});
		});

		it('supports cjs', () => {
			artifact.builds.push({ format: 'cjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						import: './cjs/index-wrapper.mjs',
						require: './cjs/index.cjs',
					},
				},
			});
		});

		it('supports cjs with types', () => {
			artifact.builds.push({ declaration: true, format: 'cjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: './cjs/index.d.ts',
						import: './cjs/index-wrapper.mjs',
						require: './cjs/index.cjs',
					},
				},
			});
		});

		it('supports .d.cts', () => {
			artifact.builds.push({ declaration: true, format: 'cjs' });
			artifact.inputs = { index: 'src/index.cts' };

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: './cjs/index.d.cts',
						import: './cjs/index-wrapper.mjs',
						require: './cjs/index.cjs',
					},
				},
			});
		});

		it('supports mjs', () => {
			artifact.builds.push({ format: 'mjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						import: './mjs/index.mjs',
					},
				},
			});
		});

		it('supports mjs with types', () => {
			artifact.builds.push({ declaration: true, format: 'mjs' });

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: './mjs/index.d.ts',
						import: './mjs/index.mjs',
					},
				},
			});
		});

		it('supports .d.mts', () => {
			artifact.builds.push({ declaration: true, format: 'mjs' });
			artifact.inputs = { index: 'src/index.mts' };

			expect(artifact.getPackageExports({})).toEqual({
				'.': {
					node: {
						types: './mjs/index.d.mts',
						import: './mjs/index.mjs',
					},
				},
			});
		});

		describe('solid', () => {
			it('adds entry point to source code', () => {
				artifact.builds.push({ format: 'lib' });

				expect(artifact.getPackageExports({ solid: true })).toEqual({
					'.': {
						node: {
							types: undefined,
							default: './lib/index.js',
						},
						solid: './src/index.ts',
						default: './lib/index.js',
					},
				});
			});

			it('supports non-bundle', () => {
				artifact.api = 'public';
				artifact.bundle = false;
				artifact.builds.push({ format: 'lib' });

				expect(artifact.getPackageExports({ solid: true })).toEqual({
					'.': {
						node: {
							types: undefined,
							default: './lib/index.js',
						},
						solid: './src/index.ts',
						default: './lib/index.js',
					},
					'./*': {
						node: {
							types: undefined,
							default: './lib/*.js',
						},
						solid: './src/*.js',
						default: './lib/*.js',
					},
				});
			});

			it('supports shared lib', () => {
				artifact.sharedLib = true;
				artifact.builds.push({ format: 'lib' });

				expect(artifact.getPackageExports({ solid: true })).toEqual({
					'.': {
						node: {
							types: undefined,
							default: './lib/node/index.js',
						},
						solid: './src/index.ts',
						default: './lib/node/index.js',
					},
				});
			});

			it('supports types', () => {
				artifact.platform = 'browser';
				artifact.builds.push({ declaration: true, format: 'esm' });

				expect(artifact.getPackageExports({ solid: true })).toEqual({
					'.': {
						browser: {
							default: undefined,
							import: './esm/index.js',
							module: './esm/index.js',
							types: './esm/index.d.ts',
						},
						solid: './src/index.ts',
					},
				});
			});
		});
	});

	describe('logWithSource()', () => {
		it('logs a message to level', () => {
			const spy = jest.spyOn(console, 'info').mockImplementation();

			artifact.log('Hello', 'info');

			expect(spy).toHaveBeenCalledWith('[project] Hello');

			spy.mockRestore();
		});

		it('includes output name and ID', () => {
			const spy = jest.spyOn(console, 'error').mockImplementation();

			artifact.log('Hello', 'error', { id: 'id', output: 'index' });

			expect(spy).toHaveBeenCalledWith(`[project:index] Hello${applyStyle(' (id=id)', 'muted')}`);

			spy.mockRestore();
		});

		it('includes source information', () => {
			const spy = jest.spyOn(console, 'warn').mockImplementation();

			artifact.log('Hello', 'warn', {
				sourceFile: fixturePath.append('test.js').path(),
				sourceLine: 10,
				sourceColumn: 55,
			});

			expect(spy).toHaveBeenCalledWith(
				`[project] Hello${applyStyle(' (file=test.js line=10:55)', 'muted')}`,
			);

			spy.mockRestore();
		});

		it('includes source line without column', () => {
			const spy = jest.spyOn(console, 'warn').mockImplementation();

			artifact.log('Hello', 'warn', {
				sourceLine: 10,
			});

			expect(spy).toHaveBeenCalledWith(`[project] Hello${applyStyle(' (line=10:?)', 'muted')}`);

			spy.mockRestore();
		});

		it('includes source column without line', () => {
			const spy = jest.spyOn(console, 'info').mockImplementation();

			artifact.log('Hello', 'info', {
				sourceColumn: 55,
			});

			expect(spy).toHaveBeenCalledWith(`[project] Hello${applyStyle(' (line=?:55)', 'muted')}`);

			spy.mockRestore();
		});
	});
});
