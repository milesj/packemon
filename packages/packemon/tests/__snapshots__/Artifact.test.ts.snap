// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Artifact > getPackageExports() > adds exports based on input file and output name 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/index.js",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > adds exports based on input file and output name when shared lib required 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/node/index.js",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > changes export namespace to "browser" when a \`browser\` platform 1`] = `
{
  ".": {
    "browser": {
      "default": "./lib/index.js",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > changes export namespace to "react-native" when a \`native\` platform 1`] = `
{
  ".": {
    "react-native": {
      "default": "./lib/index.js",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > skips \`default\` export when there is no \`lib\` build 1`] = `
{
  "./sub": {
    "node": {
      "import": {
        "default": "./mjs/sub.mjs",
        "types": undefined,
      },
      "require": {
        "default": "./cjs/sub.cjs",
        "types": undefined,
      },
    },
  },
}
`;

exports[`Artifact > getPackageExports() > solid > adds entry point to source code 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/index.js",
      "types": undefined,
    },
    "solid": "./src/index.ts",
  },
}
`;

exports[`Artifact > getPackageExports() > solid > supports non-bundle 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/index.js",
      "types": undefined,
    },
    "solid": "./src/index.ts",
  },
  "./*": {
    "node": {
      "default": "./lib/*.js",
      "types": undefined,
    },
    "solid": "./src/*.js",
  },
}
`;

exports[`Artifact > getPackageExports() > solid > supports shared lib 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/node/index.js",
      "types": undefined,
    },
    "solid": "./src/index.ts",
  },
}
`;

exports[`Artifact > getPackageExports() > solid > supports types 1`] = `
{
  ".": {
    "browser": {
      "import": "./esm/index.js",
      "types": "./esm/index.d.ts",
    },
    "solid": "./src/index.ts",
  },
}
`;

exports[`Artifact > getPackageExports() > supports .d.cts 1`] = `
{
  ".": {
    "node": {
      "default": "./cjs/index.cjs",
      "import": "./cjs/index-wrapper.mjs",
      "types": "./cjs/index.d.cts",
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports .d.mts 1`] = `
{
  ".": {
    "node": {
      "import": "./mjs/index.mjs",
      "types": "./mjs/index.d.mts",
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports cjs 1`] = `
{
  ".": {
    "node": {
      "default": "./cjs/index.cjs",
      "import": "./cjs/index-wrapper.mjs",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports cjs with types 1`] = `
{
  ".": {
    "node": {
      "default": "./cjs/index.cjs",
      "import": "./cjs/index-wrapper.mjs",
      "types": "./cjs/index.d.ts",
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports conditional exports when there are multiple builds 1`] = `
{
  ".": {
    "node": {
      "import": {
        "default": "./mjs/index.mjs",
        "types": undefined,
      },
      "require": {
        "default": "./cjs/index.cjs",
        "types": undefined,
      },
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports conditional exports with types when there are multiple builds 1`] = `
{
  ".": {
    "node": {
      "import": {
        "default": "./mjs/index.mjs",
        "types": "./mjs/index.d.ts",
      },
      "require": {
        "default": "./cjs/index.cjs",
        "types": "./cjs/index.d.ts",
      },
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports lib 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/index.js",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports lib with types 1`] = `
{
  ".": {
    "node": {
      "default": "./lib/index.js",
      "types": "./lib/index.d.ts",
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports mjs 1`] = `
{
  ".": {
    "node": {
      "import": "./mjs/index.mjs",
      "types": undefined,
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports mjs with types 1`] = `
{
  ".": {
    "node": {
      "import": "./mjs/index.mjs",
      "types": "./mjs/index.d.ts",
    },
  },
}
`;

exports[`Artifact > getPackageExports() > supports subpath file exports when output name is not "index" 1`] = `
{
  "./sub": {
    "node": {
      "default": "./lib/sub.js",
      "types": undefined,
    },
  },
}
`;
